libs/core/langchain_core/caches.py

```diff
@@ -144,16 +145,27 @@ async def aclear(self, **kwargs: Any) -> None:
 class InMemoryCache(BaseCache):
     """Cache that stores things in memory."""
 
-    def __init__(self) -> None:
-        """Initialize with empty cache."""
+    def __init__(self, maxsize: Optional[int] = None) -> None:
+        """Initialize with empty cache.
+
+        Args:
+            maxsize: The maximum number of items to store in the cache.
+                If None, the cache has no maximum size.
+                If the cache exceeds the maximum size, the oldest items are removed.
+        """
         self._cache: Dict[Tuple[str, str], RETURN_VAL_TYPE] = {}
+        if maxsize is not None and maxsize <= 0:
+            raise ValueError("maxsize must be greater than 0")
+        self._maxsize = maxsize
 
     def lookup(self, prompt: str, llm_string: str) -> Optional[RETURN_VAL_TYPE]:
         """Look up based on prompt and llm_string."""
         return self._cache.get((prompt, llm_string), None)
 
     def update(self, prompt: str, llm_string: str, return_val: RETURN_VAL_TYPE) -> None:
         """Update cache based on prompt and llm_string."""
+        if self._maxsize is not None and len(self._cache) >= self._maxsize:
```

Comment:
should this be `while`?

Reply:
Why `while`? One item is inserted and one is deleted, no?

Reply:
Then the check should be `==` not `>=` ?

Reply:
Fixed!